src/gc.c

```diff
@@ -2277,10 +2435,51 @@ JL_DLLEXPORT void jl_gc_collect(int full)
 #endif
     if (recollect) {
         n_pause--;
-        jl_gc_collect(0);
+        _jl_gc_collect(0, stack_hi);
     }
 }
 
+JL_DLLEXPORT void jl_gc_collect(int full)
+{
+    if (!is_gc_enabled || jl_in_gc)
+        return;
```

Comment:
It could be true in finalizer.

> it should perhaps be renamed `jl_in_finalizers`

Yeah, that's effectively what it is for now.